File: overlay/python/hydra-kernel/src/main/python/hydra/overlay/python/dsl/meta/defs.py
"""Meta-DSL helper for verifying module def-registration completeness.

Mirrors the role of Java's Defs.checkComplete (overlay/java/hydra-kernel/.../dsl/meta/Defs.java):
each Hydra definition in a Python source module is authored as a zero-argument, underscore-prefixed
module-level function (e.g. `def _fooBar(): return _def("fooBar")...`), then separately registered by
name in the module's definitions tuple (built by `_build_module()` or an inline list literal). Omitting
a def-producing function from that tuple silently drops it from the generated module. check_complete
catches this by calling every def-producing function found in the module and confirming its name
appears among the registered definitions.
"""

import inspect

from hydra.typed import TypedBinding


def _definition_name(d):
    """Extract the fully-qualified Name from a Definition (or a bare TypedBinding)."""
    if isinstance(d, TypedBinding):
        return d.name
    value = getattr(d, "value", None)
    return getattr(value, "name", None)


def _is_def_producer_candidate(fn):
    """A def-producing function is zero-argument and locally defined (not imported)."""
    try:
        params = inspect.signature(fn).parameters
    except (TypeError, ValueError):
        return False
    return len(params) == 0


def check_complete(module, registered):
    """Verify every def-producing function declared in `module` is present in `registered`.

    `registered` is the definitions tuple/list actually assembled for the module (e.g.
    `module_.definitions`). Walks every module-level, zero-argument function whose name starts with a
    single underscore, calls it, and checks whether the resulting definition's name is among
    `registered`'s names. A call result that isn't Definition/TypedBinding-shaped (i.e. has no
    extractable name) is treated as "not a def-producer" and skipped — not every `_foo` helper
    produces a definition. Raises AssertionError naming any orphaned function.
    """
    registered_names = {_definition_name(d) for d in registered}
    registered_names.discard(None)

    orphans = []
    for attr_name, fn in inspect.getmembers(module, inspect.isfunction):
        if fn.__module__ != module.__name__:
            continue
        if not attr_name.startswith("_") or attr_name.startswith("__"):
            continue
        if not _is_def_producer_candidate(fn):
            continue
        result = fn()
        name = _definition_name(result)
        if name is None:
            continue
        if name not in registered_names:
            orphans.append(attr_name)

    if orphans:
        raise AssertionError(
            f"check_complete: {module.__name__} declares def-producing function(s) missing from "
            f"its definitions list: {', '.join(sorted(orphans))}. Add each to the module's "
            f"definitions tuple, or remove the unused function."
        )

File: overlay/scala/hydra-kernel/src/main/scala/hydra/overlay/scala/dsl/meta/Defs.scala
package hydra.overlay.scala.dsl.meta

import hydra.packaging.Definition

/**
 * Meta-DSL helper for verifying module def-registration completeness.
 *
 * Mirrors the role of Java's Defs.checkComplete (overlay/java/hydra-kernel/.../dsl/meta/Defs.java)
 * and Python's check_complete (overlay/python/hydra-kernel/.../dsl/meta/defs.py): each Hydra
 * definition in a Scala source module is authored as a `lazy val fooDef: Definition`, then
 * separately registered in the module's `DEFINITIONS: Seq[Definition]` list. Omitting a def from
 * that list silently drops it from the generated module. checkComplete catches this via JVM
 * reflection: it finds every zero-arg, `Definition`-returning method declared on the module
 * object's class, invokes it, and confirms the result is present in `registered`.
 *
 * Unlike Java, Scala has no dedicated `Def` wrapper type to filter fields by (see Helpers.scala's
 * and Phantoms.scala's doc comments: `lazy val` replaces that need) — every def-producing binding
 * is directly typed `Definition`, so the filter is by declared/return type rather than by a marker
 * class. Unlike Python (module-level function introspection), Scala's unit of authoring is an
 * `object`'s `lazy val` members, reflected via the object's own class.
 */
object Defs:

  /**
   * Verify every zero-arg, Definition-returning method declared on `moduleObject`'s class is
   * present in `registered`.
   *
   * `moduleObject` is the source module (e.g. `Utils`, passed as `Utils` — a Scala `object`
   * reference, whose runtime class is the singleton class reflection operates on).
   * `registered` is the module's own assembly list (its `DEFINITIONS`). Raises
   * `IllegalStateException` naming any orphaned method.
   */
  def checkComplete(moduleObject: AnyRef, registered: Seq[Definition]): Unit =
    val registeredSet: Set[Definition] = registered.toSet
    val definitionClass = classOf[Definition]

    val orphans = moduleObject.getClass.getDeclaredMethods.toSeq
      .filter(m => m.getParameterCount == 0 && definitionClass.isAssignableFrom(m.getReturnType))
      .filter(m => !(m.getName == "DEFINITIONS" || m.getName == "module_"))
      .flatMap { m =>
        m.setAccessible(true)
        m.invoke(moduleObject) match
          case d: Definition if !registeredSet.contains(d) => Some(m.getName)
          case _ => None
      }
      .sorted

    if orphans.nonEmpty then
      throw new IllegalStateException(
        s"checkComplete: ${moduleObject.getClass.getSimpleName.stripSuffix("$")} declares " +
          s"def-producing member(s) missing from its DEFINITIONS list: ${orphans.mkString(", ")}. " +
          "Add each to the module's DEFINITIONS list, or remove the unused member.")
